File: integrations/alga-agent-sdk-js/src/sse.ts
import { MessageDedup } from "./dedup.js";
import { AlgaAuthError, AlgaAPIError, AlgaConnectionError } from "./errors.js";
import type {
  ConnectedEvent,
  MessageEvent as AlgaMessageEvent,
  TypingEvent,
  InvestigationSignalEvent,
  PeerFindingEvent,
  PeerAskEvent,
  PeerReplyEvent,
  CoordinationTaskEvent,
  SummarizeIncidentEvent,
  AlertAutoResolvedEvent,
  IncidentCommsStaleEvent,
} from "./models.js";

type Callback<T> = (data: T) => void;

const LOCK_EMOJI = "\u{1F512}"; // 🔒

// ErrHandler is invoked once with a terminal error (auth failure) after the
// SSE + heartbeat loops have stopped. The client must obtain a valid token and
// call start() again to resume.
export type ErrHandler = (err: Error) => void;

export interface SSEOptions {
  heartbeatIntervalMs?: number;
  userAgent?: string;
  fetchImpl?: typeof fetch;
}

export class SSEClient {
  private httpBase: string;
  private token: string;
  private dedup: MessageDedup;
  private abortController: AbortController | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectDelayMs = 2000;
  private readonly maxReconnectDelayMs = 60_000;
  private stopped = false;
  private heartbeatIntervalMs: number;
  private userAgent: string;
  private fetchImpl: typeof fetch;
  private errHandler: ErrHandler | null = null;

  onConnected: Callback<ConnectedEvent> | null = null;
  onMessage: Callback<AlgaMessageEvent> | null = null;
  onTyping: Callback<TypingEvent> | null = null;
  onInvestigationResume: Callback<InvestigationSignalEvent> | null = null;
  onPeerFinding: Callback<PeerFindingEvent> | null = null;
  onPeerAsk: Callback<PeerAskEvent> | null = null;
  onPeerReply: Callback<PeerReplyEvent> | null = null;
  onCoordinationTask: Callback<CoordinationTaskEvent> | null = null;
  onSummarizeIncident: Callback<SummarizeIncidentEvent> | null = null;
  onAlertAutoResolved: Callback<AlertAutoResolvedEvent> | null = null;
  onIncidentCommsStale: Callback<IncidentCommsStaleEvent> | null = null;
  // onUnknownEvent receives any SSE event type the SDK has no dedicated
  // handler for, so consumers can react to backend additions without an SDK
  // upgrade.
  onUnknownEvent: ((eventType: string, data: string) => void) | null = null;

  constructor(
    httpBase: string,
    token: string,
    dedup?: MessageDedup,
    heartbeatIntervalMs?: number,
    opts?: SSEOptions,
  ) {
    this.httpBase = httpBase;
    this.token = token;
    this.dedup = dedup ?? new MessageDedup();
    this.heartbeatIntervalMs = Math.max(1_000, heartbeatIntervalMs ?? 30_000);
    this.userAgent = opts?.userAgent ?? "alga-agent-sdk-js";
    this.fetchImpl = opts?.fetchImpl ?? fetch;
  }

  setErrHandler(handler: ErrHandler): void {
    this.errHandler = handler;
  }

  start(): void {
    this.stopped = false;
    this.connect();
    this.heartbeatTimer = setInterval(() => {
      this.sendHeartbeat().catch(() => {});
    }, this.heartbeatIntervalMs);
  }

  stop(): void {
    this.stopped = true;
    if (this.heartbeatTimer !== null) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
    if (this.abortController !== null) {
      this.abortController.abort();
      this.abortController = null;
    }
  }

  private fatal(err: Error): void {
    this.stopped = true;
    if (this.heartbeatTimer !== null) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
    if (this.abortController !== null) {
      this.abortController.abort();
      this.abortController = null;
    }
    this.errHandler?.(err);
  }

  private connect(): void {
    if (this.stopped) return;

    const url = `${this.httpBase}/api/v1/agent/events`;
    const ac = new AbortController();
    this.abortController = ac;

    this.fetchImpl(url, {
      headers: {
        Authorization: `Bearer ${this.token}`,
        Accept: "text/event-stream",
        "User-Agent": this.userAgent,
      },
      signal: ac.signal,
    })
      .then(async (response) => {
        if (response.status === 401 || response.status === 403) {
          throw new AlgaAuthError(response.status, "authentication failed");
        }
        if (response.status !== 200) {
          throw new AlgaAPIError(
            response.status,
            "unexpected status code",
            parseRetryAfterMs(response.headers.get("Retry-After")),
          );
        }
        if (!response.body) {
          throw new AlgaConnectionError("response has no body");
        }

        // Any successful connection resets the backoff.
        this.reconnectDelayMs = 2_000;

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";
        let eventType = "";
        let dataBuf: string[] = [];

        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done || this.stopped) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split("\n");
            buffer = lines.pop() || "";

            for (const line of lines) {
              const trimmed = line.replace(/\r$/, "");

              if (trimmed === "") {
                if (dataBuf.length > 0) {
                  this.dispatch(eventType || "message", dataBuf.join("\n"));
                }
                eventType = "";
                dataBuf = [];
                continue;
              }

              if (trimmed.startsWith(":")) continue;

              if (trimmed.startsWith("event:")) {
                eventType = trimmed.slice(6).trimStart();
                continue;
              }

              if (trimmed.startsWith("data:")) {
                // Per SSE spec, strip exactly one leading space when present.
                dataBuf.push(trimmed.slice(trimmed.charAt(5) === " " ? 6 : 5));
                continue;
              }

              // `id:` is intentionally ignored.
              if (trimmed.startsWith("id:")) continue;
            }
          }
        } catch (err: unknown) {
          if (this.stopped || ac.signal.aborted) return;
          throw new AlgaConnectionError(`stream read failed: ${(err as Error).message}`);
        }

        // Stream closed cleanly — treat as a connection error so the outer
        // loop applies backoff and reconnects.
        if (!this.stopped) throw new AlgaConnectionError("sse stream closed");
      })
      .catch((err: unknown) => {
        if (this.stopped) return;
        if (err instanceof AlgaAuthError) {
          this.fatal(err);
          return;
        }
        if (err instanceof DOMException && err.name === "AbortError") return;

        let delay: number;
        if (err instanceof AlgaAPIError && err.retryAfterMs > 0) {
          delay = err.retryAfterMs;
        } else {
          const jitter = 0.9 + Math.random() * 0.2;
          delay = Math.min(this.reconnectDelayMs * jitter, this.maxReconnectDelayMs);
          this.reconnectDelayMs = Math.min(this.reconnectDelayMs * 2, this.maxReconnectDelayMs);
        }
        setTimeout(() => this.connect(), delay);
      });
  }

  private dispatch(eventType: string, data: string): void {
    const trimmed = data.trim();
    let parsed: unknown;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      return;
    }

    switch (eventType) {
      case "connected":
        this.onConnected?.(parsed as ConnectedEvent);
        break;
      case "message": {
        const msg = parsed as AlgaMessageEvent;
        const id = msg.message_id;
        if (id && this.dedup.isDuplicate(id)) return;
        // Skip internal/system messages (leading lock-emoji per backend convention).
        if (typeof msg.text === "string" && msg.text.startsWith(LOCK_EMOJI)) return;
        this.onMessage?.(msg);
        break;
      }
      case "typing":
        this.onTyping?.(parsed as TypingEvent);
        break;
      case "investigation_resume":
        this.onInvestigationResume?.(parsed as InvestigationSignalEvent);
        break;
      case "peer_finding":
        this.onPeerFinding?.(parsed as PeerFindingEvent);
        break;
      case "peer_ask":
        this.onPeerAsk?.(parsed as PeerAskEvent);
        break;
      case "peer_reply":
        this.onPeerReply?.(parsed as PeerReplyEvent);
        break;
      case "coordination_task_dispatched":
        this.onCoordinationTask?.(parsed as CoordinationTaskEvent);
        break;
      case "summarize_incident":
        this.onSummarizeIncident?.(parsed as SummarizeIncidentEvent);
        break;
      case "alert_auto_resolved":
        this.onAlertAutoResolved?.(parsed as AlertAutoResolvedEvent);
        break;
      case "incident_comms_stale":
        this.onIncidentCommsStale?.(parsed as IncidentCommsStaleEvent);
        break;
      default:
        this.onUnknownEvent?.(eventType, trimmed);
        break;
    }
  }

  private async sendHeartbeat(): Promise<void> {
    try {
      const res = await this.fetchImpl(`${this.httpBase}/api/v1/agent/heartbeat`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.token}`,
          "User-Agent": this.userAgent,
        },
      });
      if (res.status === 401 || res.status === 403) {
        this.fatal(new AlgaAuthError(res.status, "heartbeat auth failed"));
      }
    } catch {
      // Non-auth heartbeat errors are non-fatal; the SSE loop observes
      // persistent outages on its own cadence.
    }
  }
}

// parseRetryAfterMs parses the HTTP `Retry-After` header into milliseconds.
// Both delta-seconds and HTTP-date forms are accepted. Returns 0 when the
// header is absent or unparseable. Capped at 10 minutes.
export function parseRetryAfterMs(raw: string | null): number {
  if (!raw) return 0;
  const trimmed = raw.trim();
  if (trimmed === "") return 0;
  const secs = Number(trimmed);
  if (Number.isFinite(secs) && secs >= 0) {
    return Math.min(secs * 1000, 10 * 60 * 1000);
  }
  const t = Date.parse(trimmed);
  if (Number.isFinite(t)) {
    const d = t - Date.now();
    if (d > 0) return Math.min(d, 10 * 60 * 1000);
  }
  return 0;
}

File: integrations/alga-agent-sdk-py/alga_agent_sdk/sse.py
from __future__ import annotations

import asyncio
import inspect
import json
import logging
import random
from datetime import datetime, timedelta, timezone
from email.utils import parsedate_to_datetime
from typing import Any, Callable, Optional

import httpx

from .dedup import MessageDedup
from .errors import AlgaAPIError, AlgaAuthError, AlgaConnectionError
from .models import (
    AlertAutoResolvedEvent,
    ConnectedEvent,
    CoordinationTaskEvent,
    IncidentCommsStaleEvent,
    InvestigationSignalEvent,
    MessageEvent,
    PeerAskEvent,
    PeerFindingEvent,
    PeerReplyEvent,
    SummarizeIncidentEvent,
    TypingEvent,
)

ErrHandler = Callable[[Exception], None]
EventHandler = Callable[..., Any]

_RECONNECT_BASE = 2.0
_RECONNECT_MAX = 60.0
_MAX_RETRY_AFTER = timedelta(minutes=10)
_LOCK_EMOJI = "\U0001f512"

_logger = logging.getLogger("alga_agent_sdk.sse")


def parse_retry_after(raw: Optional[str]) -> timedelta:
    if not raw:
        return timedelta(0)
    trimmed = raw.strip()
    if not trimmed:
        return timedelta(0)
    try:
        secs = int(trimmed)
    except ValueError:
        secs = None
    if secs is not None:
        if secs < 0:
            return timedelta(0)
        return min(timedelta(seconds=secs), _MAX_RETRY_AFTER)
    try:
        dt = parsedate_to_datetime(trimmed)
    except (TypeError, ValueError):
        dt = None
    if dt is not None:
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        delta = (dt - datetime.now(timezone.utc)).total_seconds()
        if delta > 0:
            return min(timedelta(seconds=delta), _MAX_RETRY_AFTER)
    return timedelta(0)


class SSEClient:
    def __init__(
        self,
        http_base: str,
        token: str,
        dedup: Optional[MessageDedup] = None,
        heartbeat_interval: float = 30.0,
        user_agent: str = "alga-agent-sdk-py",
    ):
        self._http_base = http_base.rstrip("/")
        self._token = token
        self._dedup = dedup if dedup is not None else MessageDedup()
        self._heartbeat_interval = max(1.0, heartbeat_interval)
        self._user_agent = user_agent
        self._client: Optional[httpx.AsyncClient] = None
        self._tasks: list[asyncio.Task[Any]] = []
        self._stop_event = asyncio.Event()
        self._stopped = False
        self._err_handler: Optional[ErrHandler] = None

        self.on_connected: Optional[EventHandler] = None
        self.on_message: Optional[EventHandler] = None
        self.on_typing: Optional[EventHandler] = None
        self.on_investigation_resume: Optional[EventHandler] = None
        self.on_peer_finding: Optional[EventHandler] = None
        self.on_peer_ask: Optional[EventHandler] = None
        self.on_peer_reply: Optional[EventHandler] = None
        self.on_coordination_task: Optional[EventHandler] = None
        self.on_summarize_incident: Optional[EventHandler] = None
        self.on_alert_auto_resolved: Optional[EventHandler] = None
        self.on_incident_comms_stale: Optional[EventHandler] = None
        self.on_unknown_event: Optional[EventHandler] = None

    def set_err_handler(self, handler: ErrHandler) -> None:
        self._err_handler = handler

    async def start(self) -> None:
        self._stopped = False
        self._stop_event.clear()
        self._client = httpx.AsyncClient(timeout=None)
        self._tasks = [
            asyncio.create_task(self._sse_loop()),
            asyncio.create_task(self._heartbeat_loop()),
        ]

    async def stop(self) -> None:
        self._stopped = True
        self._stop_event.set()
        for task in self._tasks:
            if not task.done():
                task.cancel()
        if self._tasks:
            await asyncio.gather(*self._tasks, return_exceptions=True)
        self._tasks = []
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    async def wait(self) -> None:
        if self._tasks:
            await asyncio.gather(*self._tasks, return_exceptions=True)

    async def _fatal(self, err: Exception) -> None:
        self._stopped = True
        self._stop_event.set()
        if self._err_handler is not None:
            try:
                self._err_handler(err)
            except Exception:
                _logger.exception("error handler raised")
        current = asyncio.current_task()
        for task in self._tasks:
            if task is not current and not task.done():
                task.cancel()

    async def _sse_loop(self) -> None:
        backoff = _RECONNECT_BASE
        while not self._stopped and not self._stop_event.is_set():
            connected, err = await self._connect_and_serve()
            if connected:
                backoff = _RECONNECT_BASE
            if err is not None:
                if isinstance(err, AlgaAuthError):
                    _logger.error(
                        "sse auth error, stopping reconnect loop: %s %s",
                        err.status_code,
                        err.message,
                    )
                    await self._fatal(err)
                    return
                delay = backoff
                if isinstance(err, AlgaAPIError) and err.retry_after > timedelta(0):
                    delay = err.retry_after.total_seconds()
                jitter = delay * (0.9 + random.random() * 0.2)
                _logger.warning("sse reconnecting after error: %s", err)
                try:
                    await asyncio.wait_for(self._stop_event.wait(), timeout=jitter)
                    return
                except asyncio.TimeoutError:
                    pass
                backoff = min(backoff * 2, _RECONNECT_MAX)
            if self._stopped or self._stop_event.is_set():
                return

    async def _connect_and_serve(self) -> tuple[bool, Optional[Exception]]:
        if self._client is None:
            return False, AlgaConnectionError("sse client not started")
        url = f"{self._http_base}/api/v1/agent/events"
        headers = {
            "Accept": "text/event-stream",
            "Authorization": f"Bearer {self._token}",
            "User-Agent": self._user_agent,
        }
        try:
            async with self._client.stream("GET", url, headers=headers) as resp:
                if resp.status_code in (401, 403):
                    return False, AlgaAuthError(
                        resp.status_code, "authentication failed"
                    )
                if resp.status_code != 200:
                    retry_after = parse_retry_after(resp.headers.get("Retry-After"))
                    return False, AlgaAPIError(
                        resp.status_code, "unexpected status code", retry_after
                    )

                event_type = ""
                data_buf: list[str] = []

                async for raw_line in resp.aiter_lines():
                    if self._stopped or self._stop_event.is_set():
                        return True, None
                    line = raw_line.rstrip("\r")

                    if line == "":
                        if data_buf:
                            ev = event_type or "message"
                            await self._dispatch(ev, "\n".join(data_buf))
                        event_type = ""
                        data_buf = []
                        continue

                    if line.startswith(":"):
                        continue

                    if line.startswith("event:"):
                        event_type = line[len("event:"):].lstrip()
                        continue

                    if line.startswith("data:"):
                        rest = line[len("data:"):]
                        if rest.startswith(" "):
                            rest = rest[1:]
                        data_buf.append(rest)
                        continue

                    if line.startswith("id:"):
                        continue

                return True, AlgaConnectionError("sse stream closed")
        except asyncio.CancelledError:
            raise
        except httpx.HTTPError as exc:
            return True, AlgaConnectionError(f"stream failed: {exc}")
        except Exception as exc:
            return True, AlgaConnectionError(f"stream failed: {exc}")

    async def _dispatch(self, event_type: str, data: str) -> None:
        data = data.strip()
        try:
            payload = json.loads(data)
        except (ValueError, TypeError):
            return

        if event_type == "connected":
            await self._invoke(self.on_connected, ConnectedEvent.model_validate(payload))
        elif event_type == "message":
            evt = MessageEvent.model_validate(payload)
            if evt.message_id and await self._dedup.is_duplicate(evt.message_id):
                return
            if evt.text.startswith(_LOCK_EMOJI):
                return
            await self._invoke(self.on_message, evt)
        elif event_type == "typing":
            await self._invoke(self.on_typing, TypingEvent.model_validate(payload))
        elif event_type == "investigation_resume":
            await self._invoke(
                self.on_investigation_resume,
                InvestigationSignalEvent.model_validate(payload),
            )
        elif event_type == "peer_finding":
            await self._invoke(
                self.on_peer_finding, PeerFindingEvent.model_validate(payload)
            )
        elif event_type == "peer_ask":
            await self._invoke(self.on_peer_ask, PeerAskEvent.model_validate(payload))
        elif event_type == "peer_reply":
            await self._invoke(
                self.on_peer_reply, PeerReplyEvent.model_validate(payload)
            )
        elif event_type == "coordination_task_dispatched":
            await self._invoke(
                self.on_coordination_task,
                CoordinationTaskEvent.model_validate(payload),
            )
        elif event_type == "summarize_incident":
            await self._invoke(
                self.on_summarize_incident,
                SummarizeIncidentEvent.model_validate(payload),
            )
        elif event_type == "alert_auto_resolved":
            await self._invoke(
                self.on_alert_auto_resolved,
                AlertAutoResolvedEvent.model_validate(payload),
            )
        elif event_type == "incident_comms_stale":
            await self._invoke(
                self.on_incident_comms_stale,
                IncidentCommsStaleEvent.model_validate(payload),
            )
        else:
            await self._invoke(self.on_unknown_event, event_type, data)

    async def _invoke(self, cb: Optional[EventHandler], *args: Any) -> None:
        if cb is None:
            return
        try:
            result = cb(*args)
            if inspect.isawaitable(result):
                await result
        except Exception:
            _logger.exception("SSE callback raised")

    async def _heartbeat_loop(self) -> None:
        while not self._stopped and not self._stop_event.is_set():
            try:
                await asyncio.wait_for(
                    self._stop_event.wait(), timeout=self._heartbeat_interval
                )
                return
            except asyncio.TimeoutError:
                pass
            if self._stopped or self._stop_event.is_set():
                return
            err = await self._post_heartbeat()
            if isinstance(err, AlgaAuthError):
                _logger.error("heartbeat auth failure, stopping: %s", err.status_code)
                await self._fatal(err)
                return
            if err is not None:
                _logger.warning("heartbeat failed: %s", err)

    async def _post_heartbeat(self) -> Optional[Exception]:
        if self._client is None:
            return None
        url = f"{self._http_base}/api/v1/agent/heartbeat"
        headers = {
            "Authorization": f"Bearer {self._token}",
            "User-Agent": self._user_agent,
        }
        try:
            resp = await self._client.post(url, headers=headers)
        except asyncio.CancelledError:
            raise
        except httpx.HTTPError as exc:
            return AlgaConnectionError(f"heartbeat failed: {exc}")
        except Exception as exc:
            return AlgaConnectionError(f"heartbeat failed: {exc}")
        if resp.status_code in (401, 403):
            return AlgaAuthError(resp.status_code, "heartbeat auth failed")
        if resp.status_code >= 400:
            return AlgaAPIError(resp.status_code, "heartbeat non-ok status")
        return None
